buildenv/jenkins/upstream_openjdk_tests

```diff
@@ -0,0 +1,306 @@
+    final ZonedDateTime newest
+    final Map items
+    
+    ReleaseMetadata(ZonedDateTime newest, Map items) {
+        this.newest = newest
+        this.items = items
+    }
+}
+
+// parallel build job configs
+Map parallelJobs = [:]
+
+// Only schedule GH API work on x86 linux nodes as there
+// are plenty available.
+node("ci.role.test&&hw.arch.x86&&sw.os.linux") {
+    parameters {
+        string(name:'VERSIONS', defaultValue: '8 11', description: "Space separated list of versions to check")
+        string(name:'GH_USERNAME', description: "Github API username to authenticate as")
+        string(name:'GH_TOKEN', description: "Github API OAuth personalized access token to be used for authentication")
+        booleanParam(name:'INCLUDE_EA', defaultValue: true, description: "Whether or not to include EA builds")
+        booleanParam(name:'FORCE_RETEST', defaultValue: false, description: "Whether or not to force a re-test irrespective previous runs")
+    }
+
+    // All params have sensible defaults
+    List versions = params.VERSIONS ? params.VERSIONS.split(" ") : "8 11".split(" ")
+    boolean includeEA = params.INCLUDE_EA ? params.INCLUDE_EA : true
+    boolean forceRetest = params.FORCE_RETEST ? params.FORCE_RETEST : false
+    String user = params.GH_USERNAME ? params.GH_USERNAME : "gh-user-not-set"
+    String token = params.GH_TOKEN ? params.GH_TOKEN : "gh-token-not-set"
+    String userTokenBase64 = Base64.getEncoder().encodeToString((user + ":" + token).getBytes())
+    Map releaseCtxts = [:]
+    Map latestRelInfos = [:]
+
+    TEST_JOB_NAME = "${env.JOB_NAME}"
+    File workspace = new File("${env.WORKSPACE}")
+    workspace.mkdirs()
+    versions.each { version ->
+        ReleaseRetriever retriever = new ReleaseRetriever(version, includeEA, userTokenBase64)
+        List ctxts = retriever.getReleaseContexts()
+        latestRelInfos.put(version, new LatestReleaseInfo(workspace, version, retriever.getLatestReleaseDate()))
+        releaseCtxts.put(version, ctxts)
+    }
+    dir("$env.WORKSPACE") { // use workspace relative directory
+        versions.each { version ->
+            stage("prepare OpenJDK Version: $version") {
+                LatestReleaseInfo relInfo = latestRelInfos.get(version)
+                List ctxts = releaseCtxts.get(version)
+                ctxts.each { println it.toString() }
+                if (forceRetest || relInfo.isRetestNeeded()) {
+                    // trigger tests for each context
+                    ctxts.each{ r -> 
+                        stage("prepare $r.OS-$r.ARCH") {
+                            def jdk_url = r.JDK_URL
+                            def source_url = r.SOURCE_URL
+                            def jdk_version = version
+                            def jenkins_file = r.getJenkinsFile()
+                            def jdk_jre_url = "$r.JDK_URL $r.JRE_URL"
+                            parallelJobs["Test_openjdk${version}_hs_sanity.openjdk_${r.ARCH}_${r.OS}"] = {
```

Comment:
@sophia-guo What parallel branch name do you suggest instead of this? `Test_upstream_openjdk...` perhaps?

Reply:
Ok, thanks!